### PR TITLE
add --watch option to `tele list`

### PR DESCRIPTION
**Pull request by njayp**
Signed-off-by: njayp <nickjaypowell@gmail.com>

## Description

A `WatchWorkloads` grpc already exists, i add a flag in `tele list` that uses it.

Reproducer: 
Run `telepresence list -w`
Hit `ctrl c`
Run a telepresence command. All commands will hang.

### CODE REVIEW

**Review by raphaelreyna — COMMENTED** [review]
This might cause some issues with the global `--output` flag since it buffers prints to stdout when outputing json. 
Perhaps check if that global flag is set to `json` and if so return an error or just set the local `--json` flag?

**Comment by njayp** [comment]
so i understand better, what issues would it cause?

**Comment by raphaelreyna** [comment]
> so i understand better, what issues would it cause?

When the `--output=json` flag is set, the `io.Writer` returned by `cmd.OutOrStdout()` is not backed by `os.Stdout` but by a `strings.Builder` which is then used to form the json response that is flushed to `os.Stdout` _after_ the command exits.
Basically, it prevents any sort of "streaming" output.

File: pkg/client/cli/cmd_list.go [hunk 1]
```diff
@@ -100,8 +118,17 @@ func (s *listInfo) list(cmd *cobra.Command, _ []string) error {
 		}
 	}
 
+	if len(workloads) == 0 {
```
Comment (thallgren) [inline pkg/client/cli/cmd_list.go]: Nit, but shouldn't this test be the very first thing that happens in the function?

File: pkg/client/cli/cmd_list.go [hunk 2]
```diff
@@ -69,22 +71,38 @@ func (s *listInfo) list(cmd *cobra.Command, _ []string) error {
...
+		stream, err := cs.userD.WatchWorkloads(ctx, &connector.WatchWorkloadsRequest{Namespaces: []string{s.namespace}}, grpc.MaxCallRecvMsgSize(int(maxRecSize)))
+		if err != nil {
+			return err
+		}
+		/*
+			go func() {
+				<-ctx.Done()
+				stream.CloseSend()
+			}()
+		*/
```
Comment (thallgren) [inline pkg/client/cli/cmd_list.go]: Please remove this comment, or explain why it's there.

**Review by thallgren — CHANGES_REQUESTED** [review]
Looks good overall. Perhaps add an integration test to verify the functionality?

**Comment by njayp** [comment]
> I can see how the term "snapshot" makes sense for a watcher that delivers snapshots, but it's a bit far fetched as an option for a list command.
> 
> * What's the difference when you list with or without a `--snapshot`?
> * When you use `--snapshot` you get JSON output.
> * Oh? I though I'd get that with the global `--json` flag?
> * Yeah, you will, but that's different, because <fill in explanation here>.

My thinking was `--output json` gets you what you usually get, but in a json obj. `--snapshot` gets you the full snapshot object. Both gets you the snapshot object in your json object. I can explain this in the changelog as a change

**Comment by thallgren** [comment]
@njayp "Both gets you the snapshot object in your json object" sounds like the `--snapshot` option is redundant, which I think it somehow isn't?

**Comment by njayp** [comment]
> @njayp "Both gets you the snapshot object in your json object" sounds like the `--snapshot` option is redundant, which I think it somehow isn't?

its not redundant because errors. without the `--output json` flag set, errors are not in json format

**Comment by thallgren** [comment]
Ok? And that's an distinction that is important enough to introduce this somewhat confusing flag?

**Comment by njayp** [comment]
> Ok? And that's an distinction that is important enough to introduce this somewhat confusing flag?

i need json errors for docker. what we could do is detect if `--output json` is set, and if it is, print the snapshot object in the json object. This would remove the ability to output the human-readable list in a json object, but maybe that isn't a common use-case. Is this what you would prefer?

**Comment by thallgren** [comment]
Ok, so why don't we change the `--json` flag to give us JSON errors then? @raphaelreyna would that be hard to do?

**Comment by thallgren** [comment]
What I don't want is different ways of telling a command to produce JSON output, motivated needs that our users rarely know anything about.

**Comment by njayp** [comment]
> Ok, so why don't we change the `--json` flag to give us JSON errors then? @raphaelreyna would that be hard to do?

I am against having multiple json flags floating around, imo having `--json` and `--output json` have different behavior is more confusing

**Comment by thallgren** [comment]
I totally agree. I think we should have one single way of asking for JSON output. The only viable use-case for it is when the CLI is called from another process, and when that happens, surely both output and errors must be returned as JSON.

**Comment by thallgren** [comment]
Something weird happened in PR #2602. It contains a whole slew of commits that doesn't belong to it and that affects the change-set of this PR (surely it doesn't span 32 files?). Some rebasing is needed in order to see what's included here.

**Review by thallgren — APPROVED** [review]

**Pull request merged**